packages/ui-truncate-list/src/TruncateList/index.tsx

```diff
@@ -0,0 +1,310 @@
+  private _resizeListener?: ResizeObserver
+
+  handleRef = (el: HTMLUListElement | null) => {
+    const { elementRef } = this.props
+
+    this.ref = el
+
+    if (typeof elementRef === 'function') {
+      elementRef(el)
+    }
+  }
+
+  constructor(props: TruncateListProps) {
+    super(props)
+
+    this.state = {
+      isMeasuring: false,
+      menuTriggerWidth: undefined
+    }
+  }
+
+  componentDidMount() {
+    this.props.makeStyles?.()
+
+    const { width: origWidth } = getBoundingClientRect(this.ref)
+
+    this._debouncedHandleResize = debounce(
+      this.handleResize,
+      this.props.debounce,
+      {
+        leading: true,
+        trailing: true
+      }
+    )
+
+    this._resizeListener = new ResizeObserver((entries) => {
+      for (const entry of entries) {
+        const { width } = entry.contentRect
+
+        if (origWidth !== width) {
+          this._debouncedHandleResize?.()
+        }
+      }
+    })
+
+    // On first render we only render the trigger to measure it
+    const menuTriggerWidth = this.props.fixMenuTriggerWidth
+      ? px(this.props.fixMenuTriggerWidth)
+      : this.calcMenuTriggerWidth()
+
+    this.setState({ menuTriggerWidth }, () => {
+      this._resizeListener?.observe(this.ref!)
+
+      this._debouncedHandleResize?.()
+    })
+  }
+
+  componentDidUpdate(
+    prevProps: TruncateListProps,
+    prevState: TruncateListState
+  ) {
+    this.props.makeStyles?.()
+
+    if (
+      prevProps.fixMenuTriggerWidth !== this.props.fixMenuTriggerWidth ||
+      prevProps.itemSpacing !== this.props.itemSpacing
+    ) {
+      this._debouncedHandleResize?.()
+    }
+
+    if (
+      this.ref &&
+      prevState.isMeasuring &&
+      prevState.isMeasuring !== this.state.isMeasuring
+    ) {
+      const menuTriggerWidth = this.calcMenuTriggerWidth()
+
+      if (
+        menuTriggerWidth &&
+        this.state.menuTriggerWidth !== menuTriggerWidth
+      ) {
+        this.setState({ menuTriggerWidth }, () => {
+          this._debouncedHandleResize?.()
+        })
+      }
+    }
+  }
+
+  componentWillUnmount() {
+    if (this._resizeListener) {
+      this._resizeListener.disconnect()
+    }
+
+    if (this._debouncedHandleResize) {
+      this._debouncedHandleResize.cancel()
+    }
+  }
+
+  get childrenArray() {
+    return Children.toArray(this.props.children)
+  }
+
+  get visibleChildren() {
+    const { visibleItemsCount } = this.props
+    const { isMeasuring, menuTriggerWidth } = this.state
+
+    // for the first render we need to measure the trigger width
+    if (typeof menuTriggerWidth === 'undefined') {
+      return []
+    }
+
+    const visibleChildren =
+      typeof visibleItemsCount === 'undefined'
+        ? this.childrenArray
+        : this.childrenArray.splice(0, visibleItemsCount)
+
+    return isMeasuring ? this.childrenArray : visibleChildren
+  }
+
+  get hiddenChildren() {
+    const { isMeasuring } = this.state
+
+    return isMeasuring
+      ? []
```

Comment:
could be a case where this is confusing to the users? I'd put it in the documentation that on the 1st render it can be empty